rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,401 @@
+```
+- Unpin a notification.
+```
+POST /api/notifications/unpin
+{ notification_id: string[] }
+```
+
+Writing to the same Notification storage might cause conflict on the `write`. If the storage mechanism implementation
+doesn’t make them impossible, we will have to use a single place to change notification status (via the [queue](#localrepository)).
+
+## Settings
+Provides HTTP API for Kibana UI to configure user-specific settings.
+Not provided in MVP, we might add in the future:
+- Unsubscribe from a source
+- Subscribe to a source
+
+#### Filtering
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+Filter configuration is passed to `Repository.get` method as `filter` property.
+```typescript
+interface NotificationFilter {
+  source_type?: string[];
+  group_id?: string[];
+  read?: boolean;
+}
+```
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Treat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
+But in addition to the restrictions listed in the [Drawbacks](#drawbacks) section, the following are added:
+- Each retrieval of a notification list still requires waiting for a roundtrip to UNS.
+- The list of Kibana notifications stored in an instance might be out of sync with a copy stored in UNS due to network problems.
+- Storing notifications in the Cloud deployment and their copy in UNS consumes twice the disk space.
+
+# Adoption strategy
+
+We start with introducing Kibana Notification Service as a Kibana plugin, adding more notification sources along the road.
+Integration with UNS is something that is not going to happen any time soon. The work can be divided into the following phases:
+
+### Phase I
+Kibana shows notifications created by the Newsfeed Kibana plugin in Notification flyout.
```

Comment:
Having the Kibana server pull the notifications from the remote newsfeed service won't work if the user has a network topology where Kibana can't access the newsfeed on the public internet. It's my understanding that this is a significant number of users. @afharo do you happen to know a rough percentage of users that have Kibana firewalled off from the public internet?

Reply:
@kobelb I think this is a key discussion you raised here (and I think it extends to more than the Newsfeed context): 

To answer your question, as of today: close to 70% of the clusters that reported in the last month, did so from the server at least once. We enabled server-side reporting ON by default [on 7.9.0](https://github.com/elastic/kibana/pull/68071). If we filter the stacks from that version, the percentage grows to ~90%. _Disclaimer: this is only for those clusters opted-in telemetry_

Newsfeed nowadays runs entirely on the public side: each user's browser fetches the newsfeed as they load the page (and keeps the data in the browser's memory, not even the local storage). The Kibana server is not involved in the process: it doesn't fetch/store anything. Even when the trend looks like we're heading to more _relaxed_ firewall rules, I think we _should_ consider UI-sourced notifications, and the implications they might have. In my experience, if we don't provide a common way to do so, plugins will _follow their own way_, potentially leading to security issues.

If we solve how UI-sourced notifications may work, I don't think we need to worry about the implementation of the Newsfeed fetcher.

Apart from solving the potential firewall issue, other UI-sourced notifications I can think of:
- Accumulate toasts? i.e.: Errors, "Upgrade your license", ...
- Search session (although this could be generated from the server)
- "Upgrade your browser for a better experience", "Use this feature to save you these 10 clicks you always go through" (maybe they are more related to Suggestion service though).

What do you think?

Reply:
> I think we should consider UI-sourced notifications, and the implications they might have. In my experience, if we don't provide a common way to do so, plugins will follow their own way, potentially leading to security issues.

This RFC doesn't distinguish between Kibana client-side and server-side sources.
But I would prefer not to introduce the concept of UI-sourced notifications as long as possible because of:
- the security concern. The browser environment is open, it much easier for an intruder to use API. For example, if we provided an HTTP endpoint to create a notification within the system, someone would use it to spam the users with notifications.
- proper separation of concerns. It's almost always a good idea to keep business logic out of UI: it's easier to reason about and test it. 
- coordination problem: a user can have Kibana open in several tabs (and on different devices). We have to come up with a solution to coordinate notification creation to avoid duplicate records.

Reply:
> If we filter the stacks from that version, the percentage grows to ~90%. Disclaimer: this is only for those clusters opted-in telemetry

and exclude Elastic Cloud the percentage is ~91%

Reply:
~90% is a lot better than I expected...@restrry do you think it's going to make the architecture a lot more complicated to get the last 10%?

Reply:
@kobelb It's doable if we are okay to live with API that has the potential problems outlined above https://github.com/elastic/kibana/pull/90297#discussion_r579096041

Are there other alternatives? What if we reconsider the decision to display News as part of the Notification experience? @thesmallestduck

Reply:
> Are there other alternatives? What if we reconsider the decision to display News as part of the Notification experience?

Newsfeed notifications are paused. We are going to wait for more use-cases of UI-sourced notifications.